sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/QuantileSummariesSuite.scala

```diff
@@ -58,7 +58,7 @@ class QuantileSummariesSuite extends SparkFunSuite {
     if (data.nonEmpty) {
       val approx = summary.query(quant).get
       // The rank of the approximation.
-      val rank = data.count(_ < approx) // has to be <, not <= to be exact
+      val rank = data.count(_ <= approx)
```

Comment:
@srowen @jiangxb1987 @thunterdb Not sure I did the right change, is the previous comment (has to be <, not <= to be exact) still correct? Similar changes were made in `DataFrameStatSuite`.

Reply:
This is trying to recover the rank of the element that was picked as the quantile. I think that's problematic when the input repeats the value chosen as the quantile. Consider estimating the median of [1,2,2,2,2,2,2,2,3]. If the method correctly picks 2, depending on whether you define this test as < or <=, you conclude that it picked rank 1 or 8 of 9 as the median. Any reasonable test of whether that rank is near the expected 5 will fail either way in some cases.

One reasonable fix it to actually use the average of `data.count(_ < approx)` and `data.count(_ <= approx)` as the implied rank that was chosen.

Do you know which test case failed without this change?

Reply:
Without this change, four test cases failed:
1. Merging ordered lists with epsi=0.1 and seq=increasing, compression=1000
2. Merging ordered lists with epsi=0.1 and seq=increasing, compression=10
3. Merging ordered lists with epsi=0.1 and seq=decreasing, compression=1000
4. Merging ordered lists with epsi=0.1 and seq=decreasing, compression=10

Reply:
I still tend to think we need to fix this test case differently, as it has the same potential problem but the other way.

Reply:
In one of the test case, `data.count(_ < approx)` = 39 and `data.count(_ <= approx)` = 40, so the average (39 + 40) / 2 < 40 (lower bound), the test still fails. Besides, data in the test suite is increasing/decreasing/random, so the case [1,2,2,2,2,2,2,2,3] can hardly happen.

Reply:
Or I can get the rank as follows, then the tests can pass:
```
      val minRank = data.count(_ < approx)
      val maxRank = data.count(_ <= approx)
      val rank = if (maxRank - minRank > 1) (minRank + maxRank) / 2 else maxRank
```
what do you think?

Reply:
I agreed that the rank here is not accurate, especially such case `[1,2,2,2,2,2,2,2,3]`.
Use average of `data.count(_ < approx) ` and `data.count(_ <= approx) ` looks more reasonable.

Reply:
@wzhfy that formula is asymmetric which feels wrong; it may happen to fix this but maybe would fail another future case. It would be a little more principled to round the average.

Yeah I know that [1,2,2,2,2,2,2,2,3] can't happen in this test, just illustrating a general point.

Hm, what's the case where the quantile is between 39 and 40? the input is 0-99 in that case? I don't see a test for the 40% quantile so wondering if we really do have a problem or are misunderstanding the failure.

Reply:
It queries 50% quantile with relativeError 0.1, then targetError is 0.1*100 = 10, so the expected rank should be in [40, 60].

Reply:
Yes I think rounding up the average can solve the problem.